File: ruby_practice_problems/02_netflix/find_median_age.rb
# Fetch relevant content based on age of users for specific country or region
# Need to update median age in real time
# Output median after input
# Store smaller half in one list, larger half in othe rlist

require './heap.rb'

# Insert age O(logn)
# Find median O(1)
# Space O(n) - storing all variables at once

class MedianOfAges
  def initialize
    @max_heap = MaxHeap.new()
    @min_heap = MinHeap.new()
  end

  def insert_age(num)
    if @max_heap.empty? || @max_heap.max >= num
      @max_heap.push(num)
    else
      @min_heap.push(num)
    end

    if @max_heap.size() > @min_heap.size() + 1
      max = @max_heap.max!
      @min_heap.push(max)
    elsif @max_heap.size() < @min_heap.size()
      min = @min_heap.min!
      @max_heap.push(min)
    end
  end

  def find_median
    if @max_heap.size() > @min_heap.size()
      return @max_heap.max()
    else
      return (@max_heap.max() + @min_heap.min()) / 2.0
    end
  end
end

medianAge = MedianOfAges.new()
medianAge.insert_age(22)
medianAge.insert_age(35)
puts("The recommended content will be for ages under: " + String(medianAge.find_median()))
medianAge.insert_age(30)
puts("The recommended content will be for ages under: " + String(medianAge.find_median()))
medianAge.insert_age(25)
puts("The recommended content will be for ages under: " + String(medianAge.find_median()))

File: javascript_practice_problems/01_netflix/find_median_age.js
import {MinHeap, MaxHeap} from './Heap.js'

class MedianOfAges{
  constructor(){
    maxHeap = new MaxHeap()
    minHeap = new MinHeap()
  }

  insertAge(num){
    if (maxHeap.isEmpty() || maxHeap.getMax() >= num){
      maxHeap.insert(num)
    }
    else
      minHeap.insert(num)

    if (maxHeap.getSize() > (minHeap.getSize() + 1)){
      var max = maxHeap.removeMax()
      minHeap.insert(max)
    }
    else if (maxHeap.getSize() < minHeap.getSize()) {
      var min = minHeap.removeMin()
      maxHeap.insert(min)
    }
  }

  findMedian(){
    if( maxHeap.getSize() == minHeap.getSize()){
      // we have even number of elements, take the average of middle two elements
      return maxHeap.getMax() / 2.0 + minHeap.getMin() / 2.0
    }
    // because max-heap will have one more element than the min-heap
    return maxHeap.getMax() / 1.0
  }
}

// Driver code

var medianAge = new MedianOfAges()
medianAge.insertAge(22)
medianAge.insertAge(35)
console.log("The recommended content will be for ages under: " + medianAge.findMedian())
medianAge.insertAge(30)
console.log("The recommended content will be for ages under: " + medianAge.findMedian())
medianAge.insertAge(25)
console.log("The recommended content will be for ages under: " + medianAge.findMedian())